Azure/sonic-buildimage
# Support SONiC Reproduceable Build-debian/pip/web packages

**- Why I did it**
Support SONiC Reproduceable Build, see design doc: https://github.com/Azure/SONiC/pull/684
Features:
1. Collect the version information to the folder target/versions when building any targets
    Sample commands:
    make configure PLATFORM=broadcom
    make target/sonic-aboot-broadcom.swi
2. Support to freeze the versions after build by command: make freeze
    Sample 1: Initialize the versions or rebuild the versions
    make freeze OPTIONS="-r"
    Sample 2: Freeze and merge the versions to current distribution and the architecture
    make freeze
    Sample 3: Merge the current target versions to all distributions and all architectures
    make freeze OPTIONS="-d -a"
    
    You can add the version change by: git add files/build/versions
    In most cases, simply run "make freeze" to freeze the versions.
3. Control the version in build, change the configuration file rules/config
    SONIC_VERSION_CONTROL_COMPONENTS=all
4. To upgrade the version configuration, just build any targets, then freeze your versions.
    Sample commands:
    make configure SONIC_VERSION_CONTROL_COMPONENTS=none PLATFORM=broadcom
    make SONIC_VERSION_CONTROL_COMPONENTS=none  target/sonic-aboot-broadcom.swi
    make freeze OPTIONS="-d -a"


**- How I did it**

**- How to verify it**
   1. Make the targets again
      Sample commands:
      make configure PLATFORM=broadcom
      make target/sonic-aboot-broadcom.swi
   2. Freeze the versions again
      make freeze
      You can see no version change in files/build/versions, git status files/build/versions.
   The version files are in target/versions, you can find and verify all the version changes for all the build targets.


**- Which release branch to backport (provide reason below if selected)**




**- Description for the changelog**



**- A picture of a cute animal (not mandatory but encouraged)**
